File: packages/soda-sdk/src/derive.ts
// SODA derivation — TS port of contracts/programs/soda/src/derivation.rs.
// Must match byte-for-byte. Vectors live in derive.test.ts.

import { secp256k1 } from "@noble/curves/secp256k1";
import { sha256 } from "@noble/hashes/sha2";
import { keccak_256 } from "@noble/hashes/sha3";

export const DERIVATION_DOMAIN = new TextEncoder().encode("SODA-v1");

export const ETH_SEPOLIA_CHAIN_TAG: Uint8Array = (() => {
  const tag = new Uint8Array(32);
  tag.set(new TextEncoder().encode("ethereum-sepolia"), 0);
  return tag;
})();

export function computeTweak(
  requesterProgram: Uint8Array,
  seeds: Uint8Array,
  chainTag: Uint8Array,
): Uint8Array {
  const h = sha256.create();
  h.update(DERIVATION_DOMAIN);
  h.update(requesterProgram);
  h.update(seeds);
  h.update(chainTag);
  return h.digest();
}

export function deriveForeignPk(
  groupPkCompressed: Uint8Array,
  tweak: Uint8Array,
): Uint8Array {
  const groupPoint = secp256k1.ProjectivePoint.fromHex(groupPkCompressed);
  const tweakBig = bytesToBigInt(tweak);
  if (tweakBig >= secp256k1.CURVE.n) {
    throw new Error("tweak >= curve order n");
  }
  const tweaked =
    tweakBig === 0n
      ? groupPoint
      : groupPoint.add(secp256k1.ProjectivePoint.BASE.multiply(tweakBig));
  return tweaked.toRawBytes(false);
}

export function ethAddressFromPk(uncompressedPk: Uint8Array): Uint8Array {
  return keccak_256(uncompressedPk.subarray(1)).subarray(12);
}

export function deriveEthAddress(
  groupPkCompressed: Uint8Array,
  requesterProgram: Uint8Array,
  seeds: Uint8Array,
  chainTag: Uint8Array,
): { tweak: Uint8Array; foreignPk: Uint8Array; ethAddress: Uint8Array } {
  const tweak = computeTweak(requesterProgram, seeds, chainTag);
  const foreignPk = deriveForeignPk(groupPkCompressed, tweak);
  const ethAddress = ethAddressFromPk(foreignPk);
  return { tweak, foreignPk, ethAddress };
}

export function bytesToBigInt(b: Uint8Array): bigint {
  let n = 0n;
  for (const byte of b) n = (n << 8n) | BigInt(byte);
  return n;
}

export function bigintToBe(n: bigint, len: number): Uint8Array {
  const out = new Uint8Array(len);
  for (let i = 0; i < len; i++) {
    out[len - 1 - i] = Number((n >> BigInt(i * 8)) & 0xffn);
  }
  return out;
}

File: contracts/signer/src/derive.rs
// SODA derivation, off-chain Rust port. Must match contracts/programs/soda/src/derivation.rs
// and packages/soda-sdk/src/derive.ts byte-for-byte.

use anyhow::{anyhow, Result};
use k256::elliptic_curve::sec1::{FromEncodedPoint, ToEncodedPoint};
use k256::elliptic_curve::PrimeField;
use k256::{AffinePoint, EncodedPoint, FieldBytes, ProjectivePoint, Scalar};
use sha2::{Digest, Sha256};

pub const DERIVATION_DOMAIN: &[u8] = b"SODA-v1";

pub fn compute_tweak(
    requester_program: &[u8; 32],
    seeds: &[u8],
    chain_tag: &[u8; 32],
) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(DERIVATION_DOMAIN);
    h.update(requester_program);
    h.update(seeds);
    h.update(chain_tag);
    h.finalize().into()
}

/// `foreign_pk = group_pk + tweak·G` returned in the 64-byte X||Y form (no
/// 0x04 prefix), matching what `secp256k1_recover` returns and what the
/// on-chain `SigRequest.foreign_pk_xy` stores.
pub fn derive_foreign_pk_xy(
    group_pk_compressed: &[u8; 33],
    tweak: &[u8; 32],
) -> Result<[u8; 64]> {
    let ep = EncodedPoint::from_bytes(group_pk_compressed)
        .map_err(|e| anyhow!("invalid group pk: {e}"))?;
    let group_affine_opt: Option<AffinePoint> = AffinePoint::from_encoded_point(&ep).into();
    let group_affine = group_affine_opt.ok_or_else(|| anyhow!("group pk not on curve"))?;

    let tweak_fb: FieldBytes = (*tweak).into();
    let tweak_scalar_opt: Option<Scalar> = Scalar::from_repr(tweak_fb).into();
    let tweak_scalar = tweak_scalar_opt.ok_or_else(|| anyhow!("tweak >= curve order n"))?;

    let group_proj = ProjectivePoint::from(group_affine);
    let tweaked_proj = group_proj + ProjectivePoint::GENERATOR * tweak_scalar;
    let tweaked_affine = AffinePoint::from(tweaked_proj);

    let ep_out = tweaked_affine.to_encoded_point(false);
    let bytes = ep_out.as_bytes();
    if bytes.len() != 65 {
        return Err(anyhow!("uncompressed encoded point not 65 bytes (got {})", bytes.len()));
    }
    let mut out = [0u8; 64];
    out.copy_from_slice(&bytes[1..]);
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn unhex<const N: usize>(s: &str) -> [u8; N] {
        let s: String = s.chars().filter(|c| !c.is_whitespace()).collect();
        let s = s.trim_start_matches("0x");
        assert_eq!(s.len(), 2 * N);
        let mut out = [0u8; N];
        for i in 0..N {
            out[i] = u8::from_str_radix(&s[i * 2..i * 2 + 2], 16).unwrap();
        }
        out
    }

    #[test]
    fn g_plus_g_equals_2g() {
        let g: [u8; 33] = unhex::<33>(
            "0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798",
        );
        let mut tweak = [0u8; 32];
        tweak[31] = 1;
        let xy = derive_foreign_pk_xy(&g, &tweak).unwrap();
        // 2G uncompressed[1..] = X||Y
        let expected: [u8; 64] = unhex::<64>(
            "C6047F9441ED7D6D3045406E95C07CD85C778E4B8CEF3CA7ABAC09B95C709EE5\
             1AE168FEA63DC339A3C58419466CEAEEF7F632653266D0E1236431A950CFE52A",
        );
        assert_eq!(xy, expected);
    }
}
